File: .agent-src/templates/scripts/work_engine/directives/ui_trivial/report.py
"""``report`` step — one-line delivery summary for the trivial path.

Phase 2 Step 6 of ``agents/roadmaps/road-to-product-ui-track.md``:
"Trivial path emits a one-line delivery summary, not the full report."

Rationale: the full backend report has nine sections — overkill for a
single-file, ≤5-line edit. The trivial summary captures what the
operator needs to see: which file, how many lines, what the edit did,
and the smoke verdict. Anything richer means the work was not
trivial, in which case ``apply`` should have reclassified to
``ui-improve`` and run the full UI track instead.

The step is pure and deterministic: reads :class:`DeliveryState`,
writes ``state.report``, always returns ``SUCCESS``.
"""
from __future__ import annotations

from typing import Any

from ...delivery_state import DeliveryState, Outcome, StepResult

AMBIGUITIES: tuple[dict[str, str], ...] = ()
"""Pure render \u2014 no blocked paths."""


def run(state: DeliveryState) -> StepResult:
    """Render the one-line trivial summary into ``state.report``."""
    state.report = _render(state)
    return StepResult(outcome=Outcome.SUCCESS)


def _render(state: DeliveryState) -> str:
    change = _last_trivial_change(state)
    if change is None:
        return "_(trivial UI edit \u2014 no change recorded)_"

    files = change.get("files") or []
    lines = change.get("lines_changed", 0)
    summary = (change.get("summary") or "trivial UI edit").strip()
    file_str = files[0] if len(files) == 1 else f"{len(files)} files"
    verdict = _smoke_verdict(state)
    verdict_str = f" \u2014 smoke: **{verdict}**" if verdict else ""
    return f"**Trivial edit:** {summary} (`{file_str}`, {lines} lines){verdict_str}"


def _last_trivial_change(state: DeliveryState) -> dict[str, Any] | None:
    for change in reversed(state.changes or []):
        if isinstance(change, dict) and change.get("kind") == "ui-trivial":
            return change
    return None


def _smoke_verdict(state: DeliveryState) -> str:
    tests = state.tests
    if isinstance(tests, dict):
        verdict = tests.get("verdict")
        if isinstance(verdict, str):
            return verdict
    return ""


__all__ = ["AMBIGUITIES", "run"]

File: src/agent-src/templates/scripts/work_engine/directives/ui_trivial/report.ts
/**
 * `report` step — one-line delivery summary for the trivial path.
 *
 * TypeScript twin of `directives/ui_trivial/report.py` (ADR-200 py2ts).
 * Public API names stay snake_case to mirror the Python module 1:1.
 *
 * The trivial summary captures what the operator needs: which file, how many
 * lines, what the edit did, and the smoke verdict. The step is pure and
 * deterministic: reads `DeliveryState`, writes `state.report`, always returns
 * `SUCCESS`.
 */
import {
    type Any,
    type DeliveryState,
    Outcome,
    StepResult,
} from '../../delivery_state.js';

/** Pure render — no blocked paths. */
export const AMBIGUITIES: ReadonlyArray<Record<string, string>> = [];

/**
 * Python truthiness for the values reached here: empty string / empty array /
 * empty object / 0 / false / null / undefined are falsy.
 */
function _pyTruthy(value: Any): boolean {
    if (value === null || value === undefined) {
        return false;
    }
    if (typeof value === 'string') {
        return value.length > 0;
    }
    if (typeof value === 'number') {
        return value !== 0;
    }
    if (typeof value === 'boolean') {
        return value;
    }
    if (Array.isArray(value)) {
        return value.length > 0;
    }
    if (typeof value === 'object') {
        return Object.keys(value).length > 0;
    }
    return true;
}

/** Python `str(value)` for the `{lines}` interpolation. */
function pyStr(value: Any): string {
    if (value === null || value === undefined) {
        return 'None';
    }
    if (value === true) {
        return 'True';
    }
    if (value === false) {
        return 'False';
    }
    return String(value);
}

function _isDict(value: Any): value is Record<string, Any> {
    return typeof value === 'object' && value !== null && !Array.isArray(value);
}

/** Render the one-line trivial summary into `state.report`. */
export function run(state: DeliveryState): StepResult {
    state.report = _render(state);
    return new StepResult({ outcome: Outcome.SUCCESS });
}

function _render(state: DeliveryState): string {
    const change = _last_trivial_change(state);
    if (change === null) {
        return '_(trivial UI edit — no change recorded)_';
    }

    const filesRaw = change['files'];
    const files = _pyTruthy(filesRaw) ? filesRaw : [];
    const lines = 'lines_changed' in change ? change['lines_changed'] : 0;
    const summaryRaw = change['summary'];
    const summary = (_pyTruthy(summaryRaw) ? String(summaryRaw) : 'trivial UI edit').trim();
    const file_str =
        Array.isArray(files) && files.length === 1
            ? pyStr((files as Any[])[0])
            : `${Array.isArray(files) ? files.length : 0} files`;
    const verdict = _smoke_verdict(state);
    const verdict_str = verdict ? ` — smoke: **${verdict}**` : '';
    return `**Trivial edit:** ${summary} (\`${file_str}\`, ${pyStr(lines)} lines)${verdict_str}`;
}

function _last_trivial_change(state: DeliveryState): Record<string, Any> | null {
    const changes = _pyTruthy(state.changes) ? state.changes : [];
    for (let i = changes.length - 1; i >= 0; i -= 1) {
        const change = changes[i];
        if (_isDict(change) && change['kind'] === 'ui-trivial') {
            return change;
        }
    }
    return null;
}

function _smoke_verdict(state: DeliveryState): string {
    const tests = state.tests;
    if (_isDict(tests)) {
        const verdict = tests['verdict'];
        if (typeof verdict === 'string') {
            return verdict;
        }
    }
    return '';
}
